test_delayed_replica_failover/test.py::test is flaky

https://s3.amazonaws.com/clickhouse-test-reports/0/a997cfad2b61200c81371202acf5c6382d3d57f6/integration_tests__asan__[4_4].html

https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgMQogICAgLS0gQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwCiAgICBBTkQgdGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJywgJ0JST0tFTicpCiAgICBBTkQgdG9EYXRlKGNoZWNrX3N0YXJ0X3RpbWUpID4gJzIwMjQtMDctMDEnCiAgICBBTkQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X2RlbGF5ZWRfcmVwbGljYV9mYWlsb3ZlciUnCk9SREVSIEJZIGNoZWNrX3N0YXJ0X3RpbWUgREVTQw==

```
            # At this point all replicas are stale, but the query must still go to second replicas which are the least stale ones.
            assert (
                instance_with_dist_table.query(
                    """
    SELECT sum(x) FROM distributed SETTINGS
        load_balancing='in_order',
        max_replica_delay_for_distributed_queries=1
    """
                ).strip()
                == "3"
            )
    
            # Prefer fallback_to_stale_replicas over skip_unavailable_shards
            assert (
                instance_with_dist_table.query(
                    """
    SELECT sum(x) FROM distributed SETTINGS
        load_balancing='in_order',
        skip_unavailable_shards=1,
        max_replica_delay_for_distributed_queries=1
    """
                ).strip()
                == "3"
            )
    
            # If we forbid stale replicas, the query must fail. But sometimes we must have bigger timeouts.
            for _ in range(20):
                try:
                    instance_with_dist_table.query(
                        """
    SELECT count() FROM distributed SETTINGS
        load_balancing='in_order',
        max_replica_delay_for_distributed_queries=1,
        fallback_to_stale_replicas_for_distributed_queries=0
    """
                    )
                    time.sleep(0.5)
                except:
                    break
            else:
>               raise Exception("Didn't raise when stale replicas are not allowed")
E               Exception: Didn't raise when stale replicas are not allowed

```

--- a/src/Client/ConnectionEstablisher.cpp
+++ b/src/Client/ConnectionEstablisher.cpp
@@ -1,6 +1,7 @@
 #include <Client/ConnectionEstablisher.h>
 #include <Common/quoteString.h>
 #include <Common/ProfileEvents.h>
+#include <Common/FailPoint.h>
 #include <Core/Settings.h>
 
 namespace ProfileEvents
@@ -25,6 +26,11 @@ namespace ErrorCodes
     extern const int DNS_ERROR;
     extern const int NETWORK_ERROR;
     extern const int SOCKET_TIMEOUT;
+}
+
+namespace FailPoints
+{
+    extern const char replicated_merge_tree_all_replicas_stale[];
 }
 
 ConnectionEstablisher::ConnectionEstablisher(
@@ -91,7 +97,15 @@ void ConnectionEstablisher::run(ConnectionEstablisher::TryResult & result, std::
 
         const UInt32 delay = table_status_it->second.absolute_delay;
         if (delay < max_allowed_delay)
+        {
             result.is_up_to_date = true;
+
+            fiu_do_on(FailPoints::replicated_merge_tree_all_replicas_stale,
+            {
+                result.delay = 1;
+                result.is_up_to_date = false;
+            });
+        }
         else
         {
             result.is_up_to_date = false;

--- a/src/Common/FailPoint.cpp
+++ b/src/Common/FailPoint.cpp
@@ -65,6 +65,7 @@ static struct InitFiu
     PAUSEABLE(infinite_sleep) \
     PAUSEABLE(stop_moving_part_before_swap_with_active) \
     REGULAR(slowdown_index_analysis) \
+    REGULAR(replicated_merge_tree_all_replicas_stale) \
 
 
 namespace FailPoints

--- a/tests/integration/test_delayed_replica_failover/test.py
+++ b/tests/integration/test_delayed_replica_failover/test.py
@@ -63,6 +63,9 @@ def create_tables():
 
 
 def test(started_cluster):
+    instance_with_dist_table.query(
+        "SYSTEM DISABLE FAILPOINT replicated_merge_tree_all_replicas_stale"
+    )
     with PartitionManager() as pm:
         # Hinder replication between replicas of the same shard, but leave the possibility of distributed connection.
         pm.partition_instances(node_1_1, node_1_2, port=9009)
@@ -124,6 +127,9 @@ SELECT sum(x) FROM distributed WITH TOTALS SETTINGS
         else:
             raise Exception("Connection with zookeeper was not lost")
 
+        instance_with_dist_table.query(
+            "SYSTEM ENABLE FAILPOINT replicated_merge_tree_all_replicas_stale"
+        )
         # At this point all replicas are stale, but the query must still go to second replicas which are the least stale ones.
         assert (
             instance_with_dist_table.query(

